### DISCUSSION TITLE
Mass Storage

### DISCUSSION THREAD

**Discussion by ce4aser** [micropython/micropython#16012] [General]
Hello, 

how setup up, micropython to mass storage (like circuitpython). Becouse i see files only in thonny.

Thanks for answer.

**Comment by jonnor** [comment]
USB mass storage device for access to internal files is not supported in MicroPython. Thonny, ViperIDE or mpremote is needed to access files. With mpremote I believe you can mount the device as directory.

**Reply by scruss** [reply]
It's supported on some boards, but not recommended. It's not possible on ESP8266 or most ESP32s, since they require a USB UART chip that gets in the way.

The OP didn't say what board they were using. While MSC is convenient, it's not reliable. It's the first thing I turn off if I have to use CircuitPython

**Reply by ce4aser** [reply]
> It's supported on some boards, but not recommended. It's not possible on ESP8266 or most ESP32s, since they require a USB UART chip that gets in the way.
> 
> The OP didn't say what board they were using. While MSC is convenient, it's not reliable. It's the first thing I turn off if I have to use CircuitPython

Raspberry Pi Pico, Raspberry Pi Pico W, Raspberry Pi Pico 2, Arduino Nano RP2040 Connect etc. Next week Pimoroni Pico Plus 2. So boards of RP2040 or RP2350 MCU.

Why not recommend?

**Reply by peterhinch** [reply]
> Why not recommend?

See [explanation](https://github.com/peterhinch/micropython-samples/blob/master/README.md#15-a-usb-pitfall).

**Reply by dhalbert** [reply]
Note that CircuitPython makes the filesystem read-only to Python code when it is read-write as a USB drive and vice-versa. This prevents a major source of corruption, and thus prevents some of the corruption problems mentioned in @peterhinch's [explanation](https://github.com/peterhinch/micropython-samples/blob/master/README.md#15-a-usb-pitfall).  It is still the case that FAT filesystems are relatively fragile: a reset during a write, etc. will cause corruption.

One of CircuitPython's primary motivations for keeping MSC access is immediate usability without installing extra tools, which in some cases is not possible (school environments, etc.). Browser-based tools are a newer way around that. We respect MicroPython's choices in this matter, since we serve somewhat different, if overlapping, audiences.

If only Apple had supported MTP (Media Transfer Protocol)! But it would have allowed third-party MP3 players to compete against iPods more easily on Macs, so naturally they didn't want to enable that.

**Reply by scruss** [reply]
> Note that CircuitPython makes the filesystem read-only to Python code

I still get maddening levels of FS corruption with CircuitPython until I turn MSC off, though.

**Reply by peterhinch** [reply]
Interesting. I think MSC is an abomination. Without it, I don't think I've ever experienced corruption.

**Reply by andrewleech** [reply]
Now that runtime USB device support is available I've been wanting to add MTP protocol support; this is what's used by Android phones etc to provide USB filesystem access. 

In many ways it looks identical to MSC however on the device side it's managed at the file level, not filesystem/block, so it can be supported safely by micropython for any underlying filesystem format (ie lfs will work fine).

It's got out of the box support on Windows and Linux, but unfortunately not Mac. I can only presume Apple does this out of protest against Google. There's file manager apps for Mac that fill the gap though.

So yeah lots of benefits with little to no downsides, other than finding the time to implement it :-D
